### DISCUSSION TITLE
How to convert Skype tags and symbols to smiles?

### DISCUSSION THREAD

**Discussion by mihalt** [Terrance/SkPy#230] [Q&A] [answered]
Hi! Only would like to ask you, how would you resolve this problem? I want to see smiles in my str output in admin pannel. Not only tags, that contain them. Maybe is separate library of smiles exist, that skype only is using by itself?

**Comment by Terrance** [comment, marked as answer]
If you mean resolving markup like `(sad)` into the corresponding emoticons, you probably want to take a look at [`SkypeUtils.static`](https://skpy.t.allofti.me/reference/internal.html#skpy.util.SkypeUtils.static) -- this is a `dict` of metadata about the emoticons that Skype provides.  For example:

```python
>>> tag = "(sad)"
>>> for item in SkypeUtils.static["items"]:
...   if tag in item["shortcuts"]:
...     print(item)
...
{'description': 'Sad',
 'etag': 'e34deb20-f8f5-4b53-ab00-468ccacfac69',
 'facebook': {'shortcuts': [':(']},
 'id': 'sad',
 'keywords': [],
 'media': {'default': {'firstFrame': 17, 'fps': 24, 'framesCount': 63}},
 'shortcuts': [':(', ':-(', ':=(', '(sad)', '(Sad)', ':<', ':-<', ':(', ':-(', ':=(', '(sad)', '(Sad)', ':<', ':-<'],
 'type': 'emoticon',
 'useInSms': True,
 'visible': True}
```

For the actual image, the current Skype for Web seems to use URLs of the form:
```
https://static-asm.secure.skypeassets.com/pes/v1/emoticons/<id>/views/default_<size>_f
```

`id` comes from the item metadata above, and `size` is one of a handful of recognised numbers, up to 100.

Here is `sad` at size 20: ![](https://static-asm.secure.skypeassets.com/pes/v1/emoticons/sad/views/default_20_f)

**Reply by mihalt** [reply]
> If you mean resolving markup like `(sad)` into the corresponding emoticons, you probably want to take a look at [`SkypeUtils.static`](https://skpy.t.allofti.me/reference/internal.html#skpy.util.SkypeUtils.static) -- this is a `dict` of metadata about the emoticons that Skype provides. For example:
> 
> ```python
> >>> tag = "(sad)"
> >>> for item in SkypeUtils.static["items"]:
> ...   if tag in item["shortcuts"]:
> ...     print(item)
> ...
> {'description': 'Sad',
>  'etag': 'e34deb20-f8f5-4b53-ab00-468ccacfac69',
>  'facebook': {'shortcuts': [':(']},
>  'id': 'sad',
>  'keywords': [],
>  'media': {'default': {'firstFrame': 17, 'fps': 24, 'framesCount': 63}},
>  'shortcuts': [':(', ':-(', ':=(', '(sad)', '(Sad)', ':<', ':-<', ':(', ':-(', ':=(', '(sad)', '(Sad)', ':<', ':-<'],
>  'type': 'emoticon',
>  'useInSms': True,
>  'visible': True}
> ```
> 
> For the actual image, the current Skype for Web seems to use URLs of the form:
> 
> ```
> https://static-asm.secure.skypeassets.com/pes/v1/emoticons/<id>/views/default_<size>_f
> ```
> 
> `id` comes from the item metadata above, and `size` is one of a handful of recognised numbers, up to 100.
> 
> Here is `sad` at size 20: ![](https://camo.githubusercontent.com/3d999cfe9677b63ead64d8995fb3fc68af943445190509354e8564f8d0b8dd49/68747470733a2f2f7374617469632d61736d2e7365637572652e736b7970656173736574732e636f6d2f7065732f76312f656d6f7469636f6e732f7361642f76696577732f64656661756c745f32305f66)

oh, you are my god! Thank you very much! Very wait when you will resolve v2 skype api problems with current enhancement . There are a lot of useful functionality in them. Maybe do you need any help?

**Comment by mihalt** [comment]
@Terrance I just found, that not all emotion are in SkypeUtils.static["items"]. For example this shortcut (1f380_ribbon). Do you know the link, how I can get id of emotion by it's shortcut?

**Reply by Terrance** [reply]
That appears to be a [Unicode emoji](https://emojipedia.org/ribbon/) -- I guess Skype has aliases for those now too?

If it's not in the emoticon data, you could at least decode the character codepoint to the literal emoji:

```python
>>> chr(int('1f380', 16))
'🎀'
```

**Reply by mihalt** [reply]
hmm, ok. So, should I find by pattern like `re.findall('(.*?)_(.*?)) → [(literal emoji), (skype_key)]`? Or how should it look like?

**Reply by mihalt** [reply]
@Terrance I still haven't resolved this issue. Can maybe the lib like https://github.com/carpedm20/emoji/ help too?

**Reply by Terrance** [reply]
I don't know what more you're trying to do here -- for decoding emoji-like emoticons like the example you gave above, it looks like the first component is the codepoint, so you can use `chr(int(shortcut.split("_", 0)[0], 16))` as above; you probably don't need to encode as you can just send literal emoji directly?

**Reply by mihalt** [reply]
> I don't know what more you're trying to do here -- for decoding emoji-like emoticons like the example you gave above, it looks like the first component is the codepoint, so you can use `chr(int(shortcut.split("_", 0)[0], 16))` as above; you probably don't need to encode as you can just send literal emoji directly?

I try to download all emotions images to my database and match with chars combination to dinamically change them to image, when I try to render downloaded message in my app. 

As I see from what I get from skype I have names of images like this `dance_tSLACrp`. So, the patterns that you and I provided can not work strings like that.

**Comment by mihalt** [comment]
@Terrance maybe you should know, that this ids from `SkypeUtils.static["items"]` should be changed for correct getting of image by link, that you game me `https://static-asm.secure.skypeassets.com/pes/v1/emoticons/<id>/views/default_<size>_f`

```
EMOTION_KEY_MATCH = {
    'eg': 'evilgrin',
    'bertlett': 'bartlett',
    'mlt': 'malthe',
    'taur': 'tauri'
}
```

Another are fine. 

And second maybe that you should know that `id='movember'` contains shortcut `(mo)` that is duplicated with `id='cash'`. Another shortcuts in another ids are fine

And third — the same key can contain in conversations different shortcuts. For example, when New Year holidays are comming, `id=party` contains shortcut in messages `(festiveparty)`. In standard days `id=party` shortcut is `(party)` and `id=festiveparty` is `(festiveparty)` separately.

**Reply by Terrance** [reply]
If Skype is returning bad data here then I don't intend to patch or second-guess it -- the emoticon data is provided as-is.  Ditto if the clients vary the meaning of shortcuts based on time of year.

SkPy's only use of it is through the [`SkypeMsg.emote`](https://skpy.t.allofti.me/reference/api.html#skpy.msg.SkypeMsg.emote) helper that looks up a shortcut, so if there's multiple emoticons sharing a shortcut then you'll get the first match.  If you need more control when sending then you can construct the `<ss>` tag yourself.
